File: ios/CodexMeterCore/Sources/CodexMeterCore/UsagePace.swift
import Foundation

public enum UsagePaceSensitivity: String, Codable, Sendable, CaseIterable {
    case off
    case sensitive
    case balanced
    case relaxed
}

public enum UsagePaceWindow: String, Codable, Sendable {
    case fiveHour
    case weekly
    case monthly
}

public struct UsagePaceAssessment: Sendable, Equatable {
    public let isAvailable: Bool
    public let isAccelerated: Bool
    public let estimatedRemaining: TimeInterval
    public let estimatedTotal: TimeInterval
    public let actualRemaining: TimeInterval
    public let estimatedExhaustionAt: Date?
    public let resetAt: Date?
    public let observedDuration: TimeInterval
    public let usedPercent: Int

    public var coverageRatio: Double {
        guard isAvailable, actualRemaining > 0 else { return .infinity }
        return estimatedRemaining / actualRemaining
    }

    public static let unavailable = UsagePaceAssessment(
        isAvailable: false,
        isAccelerated: false,
        estimatedRemaining: 0,
        estimatedTotal: 0,
        actualRemaining: 0,
        estimatedExhaustionAt: nil,
        resetAt: nil,
        observedDuration: 0,
        usedPercent: 0
    )
}

/// Estimates quota exhaustion from full-window usage and optional recent local history.
public enum UsagePace {
    public static func assess(
        window: UsageWindow?,
        history: UsageHistory? = nil,
        observedAt: Date,
        now: Date = Date(),
        sensitivity: UsagePaceSensitivity = .balanced
    ) -> UsagePaceAssessment {
        let baseline = baselineAssessment(
            window: window,
            observedAt: observedAt,
            now: now,
            sensitivity: sensitivity
        )
        guard baseline.isAvailable,
              let window,
              let history,
              history.observedBurnRate > 0 else {
            return baseline
        }

        let averageRate = Double(window.usedPercent) / baseline.observedDuration
        guard averageRate.isFinite, averageRate > 0 else {
            return baseline
        }
        let observedRate = min(
            averageRate * 4,
            max(averageRate * 0.25, history.observedBurnRate)
        )
        let blendedRate = averageRate * 0.4 + observedRate * 0.6
        guard blendedRate.isFinite, blendedRate > 0,
              let resetAt = baseline.resetAt else {
            return baseline
        }

        let remainingAtObservation = max(
            0,
            Double(100 - window.usedPercent) / blendedRate
        )
        let exhaustionAt = observedAt.addingTimeInterval(remainingAtObservation)
        let estimatedRemaining = max(0, exhaustionAt.timeIntervalSince(now))
        let estimatedTotal = baseline.observedDuration + remainingAtObservation
        let policy = policy(for: sensitivity, windowDuration: TimeInterval(window.windowSeconds))
        let accelerated = sensitivity != .off
            && remainingAtObservation * 100
                <= resetAt.timeIntervalSince(observedAt) * Double(policy.maximumCoveragePercent)

        return UsagePaceAssessment(
            isAvailable: true,
            isAccelerated: accelerated,
            estimatedRemaining: estimatedRemaining,
            estimatedTotal: estimatedTotal,
            actualRemaining: baseline.actualRemaining,
            estimatedExhaustionAt: exhaustionAt,
            resetAt: resetAt,
            observedDuration: baseline.observedDuration,
            usedPercent: window.usedPercent
        )
    }

    public static func mostAcceleratedWindow(
        in snapshot: UsageSnapshot?,
        histories: [UsageHistoryKind: UsageHistory] = [:],
        now: Date = Date(),
        sensitivity: UsagePaceSensitivity = .balanced
    ) -> UsagePaceWindow? {
        guard sensitivity != .off, let snapshot else { return nil }
        let fiveHour = assess(
            window: snapshot.fiveHour,
            history: histories[.fiveHour],
            observedAt: snapshot.fetchedAt,
            now: now,
            sensitivity: sensitivity
        )
        let weekly = assess(
            window: snapshot.weekly,
            history: histories[.weekly],
            observedAt: snapshot.fetchedAt,
            now: now,
            sensitivity: sensitivity
        )
        let monthly = assess(
            window: snapshot.monthly,
            history: histories[.monthly],
            observedAt: snapshot.fetchedAt,
            now: now,
            sensitivity: sensitivity
        )

        let assessments: [(UsagePaceWindow, UsagePaceAssessment)] = [
            (.fiveHour, fiveHour),
            (.weekly, weekly),
            (.monthly, monthly)
        ]
        return assessments
            .filter(\.1.isAccelerated)
            .min { $0.1.coverageRatio < $1.1.coverageRatio }?
            .0
    }

    private static func baselineAssessment(
        window: UsageWindow?,
        observedAt: Date,
        now: Date,
        sensitivity: UsagePaceSensitivity
    ) -> UsagePaceAssessment {
        guard let window,
              window.windowSeconds > 0,
              observedAt.timeIntervalSince1970.isFinite,
              observedAt.timeIntervalSince1970 > 0,
              now.timeIntervalSince1970.isFinite,
              let resetAt = window.effectiveResetDate(relativeTo: observedAt),
              resetAt > observedAt,
              resetAt > now else {
            return .unavailable
        }

        let windowDuration = TimeInterval(window.windowSeconds)
        let windowStart = resetAt.addingTimeInterval(-windowDuration)
        let elapsed = observedAt.timeIntervalSince(windowStart)
        guard elapsed.isFinite,
              elapsed > 0,
              elapsed <= windowDuration,
              window.usedPercent > 0 else {
            return .unavailable
        }

        let policy = policy(for: sensitivity, windowDuration: windowDuration)
        guard elapsed >= policy.minimumElapsed,
              window.usedPercent >= policy.minimumUsedPercent else {
            return .unavailable
        }

        let estimatedRemainingAtObservation =
            elapsed * Double(100 - window.usedPercent) / Double(window.usedPercent)
        let estimatedTotal = elapsed * 100 / Double(window.usedPercent)
        guard estimatedRemainingAtObservation.isFinite, estimatedTotal.isFinite else {
            return .unavailable
        }
        let exhaustionAt = observedAt.addingTimeInterval(estimatedRemainingAtObservation)
        let actualRemainingAtObservation = resetAt.timeIntervalSince(observedAt)
        let accelerated = sensitivity != .off
            && estimatedRemainingAtObservation * 100
                <= actualRemainingAtObservation * Double(policy.maximumCoveragePercent)

        return UsagePaceAssessment(
            isAvailable: true,
            isAccelerated: accelerated,
            estimatedRemaining: max(0, exhaustionAt.timeIntervalSince(now)),
            estimatedTotal: estimatedTotal,
            actualRemaining: max(0, resetAt.timeIntervalSince(now)),
            estimatedExhaustionAt: exhaustionAt,
            resetAt: resetAt,
            observedDuration: elapsed,
            usedPercent: window.usedPercent
        )
    }

    private static func policy(
        for sensitivity: UsagePaceSensitivity,
        windowDuration: TimeInterval
    ) -> (
        minimumUsedPercent: Int,
        minimumElapsed: TimeInterval,
        maximumCoveragePercent: Int
    ) {
        switch sensitivity {
        case .sensitive:
            (2, max(2 * 60, windowDuration / 400), 100)
        case .relaxed:
            (10, max(10 * 60, windowDuration / 100), 50)
        case .off, .balanced:
            (5, max(5 * 60, windowDuration / 200), 75)
        }
    }
}

File: android/shared/src/main/java/dev/scorpion7slayer/modelsmeter/UsagePace.java
package dev.scorpion7slayer.modelsmeter;

import java.util.concurrent.TimeUnit;

/**
 * Estimates quota exhaustion from the average consumption over the complete current window.
 *
 * <p>The elapsed sample begins at {@code resetAt - windowDuration}, so pauses and idle periods
 * remain part of the average instead of making a later burst look permanently representative.
 */
public final class UsagePace {
    public static final String OFF = "off";
    public static final String SENSITIVE = "sensitive";
    public static final String BALANCED = "balanced";
    public static final String RELAXED = "relaxed";

    public static final int WINDOW_NONE = 0;
    public static final int WINDOW_FIVE_HOUR = 1;
    public static final int WINDOW_WEEKLY = 2;
    public static final int WINDOW_MONTHLY = 3;

    private UsagePace() {
    }

    public static Assessment assess(UsageWindow window, long observedAtMillis, long nowMillis,
            String sensitivity) {
        if (window == null || window.windowSeconds <= 0L || observedAtMillis <= 0L) {
            return Assessment.unavailable();
        }
        long windowMillis = safeMultiply(window.windowSeconds, 1000L);
        long resetAtMillis = window.effectiveResetAtMillis(observedAtMillis);
        if (windowMillis <= 0L || resetAtMillis <= observedAtMillis
                || resetAtMillis <= nowMillis) {
            return Assessment.unavailable();
        }
        long windowStartMillis = resetAtMillis - windowMillis;
        long elapsedMillis = observedAtMillis - windowStartMillis;
        if (elapsedMillis <= 0L || elapsedMillis > windowMillis || window.usedPercent <= 0) {
            return Assessment.unavailable();
        }

        String normalized = normalizeSensitivity(sensitivity);
        Policy policy = policy(normalized, windowMillis);
        if (elapsedMillis < policy.minimumElapsedMillis
                || window.usedPercent < policy.minimumUsedPercent) {
            return Assessment.unavailable();
        }

        long estimatedRemainingAtObservation = safeMultiply(elapsedMillis,
                100L - window.usedPercent) / window.usedPercent;
        long estimatedTotalMillis = safeMultiply(elapsedMillis, 100L) / window.usedPercent;
        long estimatedExhaustionAtMillis = safeAdd(observedAtMillis,
                estimatedRemainingAtObservation);
        long actualRemainingAtObservation = resetAtMillis - observedAtMillis;
        boolean accelerated = warningsEnabled(normalized)
                && safeMultiply(estimatedRemainingAtObservation, 100L)
                <= safeMultiply(actualRemainingAtObservation, policy.maximumCoveragePercent);
        long estimatedRemainingNow = Math.max(0L, estimatedExhaustionAtMillis - nowMillis);
        long actualRemainingNow = Math.max(0L, resetAtMillis - nowMillis);
        return new Assessment(true, accelerated, estimatedRemainingNow, estimatedTotalMillis,
                actualRemainingNow, estimatedExhaustionAtMillis, resetAtMillis,
                elapsedMillis, window.usedPercent);
    }

    /**
     * Refines the full-window average with actual same-window samples when enough history exists.
     * The blend keeps estimates stable while reacting to a sustained recent change in burn.
     */
    public static Assessment assess(UsageWindow window, UsageHistory history,
            long observedAtMillis, long nowMillis, String sensitivity) {
        Assessment baseline = assess(window, observedAtMillis, nowMillis, sensitivity);
        if (!baseline.available || history == null) return baseline;
        double observedRate = history.observedBurnRate();
        if (observedRate <= 0d) return baseline;
        double averageRate = window.usedPercent / (double) baseline.observedDurationMillis;
        if (averageRate <= 0d) return baseline;
        // Bound malformed or unusually bursty samples before blending.
        observedRate = Math.max(averageRate * 0.25d, Math.min(averageRate * 4d, observedRate));
        double blendedRate = averageRate * 0.4d + observedRate * 0.6d;
        long remainingAtObservation = Math.max(0L,
                Math.round((100d - window.usedPercent) / blendedRate));
        long exhaustionAt = safeAdd(observedAtMillis, remainingAtObservation);
        long estimatedRemainingNow = Math.max(0L, exhaustionAt - nowMillis);
        long estimatedTotal = safeAdd(baseline.observedDurationMillis, remainingAtObservation);
        Policy policy = policy(normalizeSensitivity(sensitivity),
                safeMultiply(window.windowSeconds, 1000L));
        boolean accelerated = warningsEnabled(sensitivity)
                && safeMultiply(remainingAtObservation, 100L)
                <= safeMultiply(baseline.resetAtMillis - observedAtMillis,
                        policy.maximumCoveragePercent);
        return new Assessment(true, accelerated, estimatedRemainingNow, estimatedTotal,
                baseline.actualRemainingMillis, exhaustionAt, baseline.resetAtMillis,
                baseline.observedDurationMillis, window.usedPercent);
    }

    public static int mostAcceleratedWindow(UsageSnapshot snapshot, long nowMillis,
            String sensitivity) {
        if (snapshot == null || !warningsEnabled(sensitivity)) return WINDOW_NONE;
        Assessment[] assessments = {
                assess(snapshot.fiveHour, snapshot.fetchedAtMillis, nowMillis, sensitivity),
                assess(snapshot.weekly, snapshot.fetchedAtMillis, nowMillis, sensitivity),
                assess(snapshot.monthly, snapshot.fetchedAtMillis, nowMillis, sensitivity)
        };
        int[] windows = {WINDOW_FIVE_HOUR, WINDOW_WEEKLY, WINDOW_MONTHLY};
        int selected = WINDOW_NONE;
        double selectedRatio = Double.POSITIVE_INFINITY;
        for (int i = 0; i < assessments.length; i++) {
            if (!assessments[i].accelerated) continue;
            double ratio = assessments[i].coverageRatio();
            if (selected == WINDOW_NONE || ratio < selectedRatio) {
                selected = windows[i];
                selectedRatio = ratio;
            }
        }
        return selected;
    }

    public static String normalizeSensitivity(String sensitivity) {
        if (OFF.equals(sensitivity) || SENSITIVE.equals(sensitivity)
                || RELAXED.equals(sensitivity)) {
            return sensitivity;
        }
        return BALANCED;
    }

    /** Whether accelerated-usage warnings (and their Now Bar triggers) may fire. */
    public static boolean warningsEnabled(String sensitivity) {
        return !OFF.equals(normalizeSensitivity(sensitivity));
    }

    private static Policy policy(String sensitivity, long windowMillis) {
        switch (normalizeSensitivity(sensitivity)) {
            case SENSITIVE:
                return new Policy(2, Math.max(TimeUnit.MINUTES.toMillis(2),
                        windowMillis / 400L), 100);
            case RELAXED:
                return new Policy(10, Math.max(TimeUnit.MINUTES.toMillis(10),
                        windowMillis / 100L), 50);
            case OFF:
                // Keep balanced sample gates so estimates still appear without warnings.
            default:
                return new Policy(5, Math.max(TimeUnit.MINUTES.toMillis(5),
                        windowMillis / 200L), 75);
        }
    }

    private static long safeMultiply(long left, long right) {
        if (left <= 0L || right <= 0L) return 0L;
        if (left > Long.MAX_VALUE / right) return Long.MAX_VALUE;
        return left * right;
    }

    private static long safeAdd(long left, long right) {
        if (right > 0L && left > Long.MAX_VALUE - right) return Long.MAX_VALUE;
        return left + right;
    }

    private static final class Policy {
        final int minimumUsedPercent;
        final long minimumElapsedMillis;
        final int maximumCoveragePercent;

        Policy(int minimumUsedPercent, long minimumElapsedMillis,
                int maximumCoveragePercent) {
            this.minimumUsedPercent = minimumUsedPercent;
            this.minimumElapsedMillis = minimumElapsedMillis;
            this.maximumCoveragePercent = maximumCoveragePercent;
        }
    }

    public static final class Assessment {
        public final boolean available;
        public final boolean accelerated;
        public final long estimatedRemainingMillis;
        public final long estimatedTotalMillis;
        public final long actualRemainingMillis;
        public final long estimatedExhaustionAtMillis;
        public final long resetAtMillis;
        public final long observedDurationMillis;
        public final int usedPercent;

        private Assessment(boolean available, boolean accelerated,
                long estimatedRemainingMillis, long estimatedTotalMillis,
                long actualRemainingMillis, long estimatedExhaustionAtMillis,
                long resetAtMillis, long observedDurationMillis, int usedPercent) {
            this.available = available;
            this.accelerated = accelerated;
            this.estimatedRemainingMillis = estimatedRemainingMillis;
            this.estimatedTotalMillis = estimatedTotalMillis;
            this.actualRemainingMillis = actualRemainingMillis;
            this.estimatedExhaustionAtMillis = estimatedExhaustionAtMillis;
            this.resetAtMillis = resetAtMillis;
            this.observedDurationMillis = observedDurationMillis;
            this.usedPercent = usedPercent;
        }

        private static Assessment unavailable() {
            return new Assessment(false, false, 0L, 0L, 0L, 0L, 0L, 0L, 0);
        }

        double coverageRatio() {
            if (!available || actualRemainingMillis <= 0L) return Double.POSITIVE_INFINITY;
            return (double) estimatedRemainingMillis / (double) actualRemainingMillis;
        }
    }
}
